### DISCUSSION TITLE
learning Chaboche modele implementation

### DISCUSSION THREAD

**Discussion by GhiathMonnet** [thelfer/tfel#397] [General]
Dear All,
using the Chaboche model implementation (see below), I had two questions : 
- is there any documentation on the function computeElasticPrediction() ? I could not find it. The line command help system requires the @ symbol.
- in this implementation, which seems to work well, the elastic strain is updated by :  (feel += dp*n ;). This looks as a plastic correction of the trial prediction. This means that eel has already updated by adding the deto increment. But when looking carefully into the file, I did not see any line assigning deto to eel. A mystery for me.
Many thanks for help.
Ghiath
The implementation I am looking at is obtained by the link from the mfront site. Below the corresponding file : 

@DSL       Implicit;
@Author    Jean-Michel Proix;
@Date      15/04/2014;
@Behaviour ChabocheViscoplasticity;
@Brick     StandardElasticity;

@Theta 1.;
@Epsilon 1.e-13;

@MaterialProperty stress young;
young.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");

@StateVariable strain p;
p.setGlossaryName("EquivalentViscoplasticStrain");
@StateVariable StrainStensor a[2];

@MaterialProperty stress Rinf,R0;
@MaterialProperty real b;
@MaterialProperty stress C[2];
@MaterialProperty real   g[2];
@MaterialProperty real m;
@MaterialProperty real K;

@LocalVariable bool bloading;

@InitLocalVariables{
  StressStensor scin = computeElasticPrediction();
  for(unsigned short i=0;i!=2;++i){
     scin  -= 2*C[i]*a[i]/3;
  }
  const stress seq = sigmaeq(scin);
  const stress Rp  = Rinf + (R0-Rinf)*exp(-b*p);  
  bloading = seq - Rp > 0;
}

@Integrator{
  constexpr const real eps = 1.e-12;
  if(bloading){
    const auto mu      = computeMu(young,nu);
    const strain p_    = p +theta*dp ;
    StressStensor scin = sig;
    StrainStensor a_[2];
    for(unsigned short i=0;i!=2;++i){
      a_[i]  = a[i]+theta*da[i];
      scin  -= 2*C[i]*a_[i]/3;
    }
    const stress seq = sigmaeq(scin);
    const stress Rp  = Rinf + (R0-Rinf)*exp(-b*p_) ;  
    const stress F = max(seq - Rp,stress(0));
    const real Fexp   = pow(F/K,m-1)/K;
    const auto iseq = 1/max(seq,eps*young);
    const Stensor n = 3*deviator(scin)*iseq/2;
    const auto vp  = Fexp*F;
    feel += dp*n;
    fp   -= vp*dt;
    for(unsigned short i=0;i!=2;++i){
      fa[i] = da[i] - dp*(n-g[i]*a_[i]);
    }
    const Stensor4 Jmn  = Stensor4::M() - (n^n);    
    dfeel_ddeel += 2*mu*theta*dp*Jmn*iseq ;
    dfeel_ddp    = n;
    dfp_ddeel    = - 2*Fexp*m*dt* 2.*mu*theta*(n|Stensor4::M())/3;
    dfp_ddp      = 1 + theta* Fexp *m *dt*b*(Rinf-Rp);
    for(unsigned short i=0;i!=2;++i){
      dfeel_dda(i)   = -2*C[i]*dp*theta* iseq/3*Jmn ;
      dfp_dda(i)     =  2*Fexp*m*dt*C[i]*theta/3*n;  
      dfa_ddeel(i)   = -2*mu*theta*dp*Jmn *iseq;
      dfa_ddp(i)     = -n + g[i]*a_[i];
      dfa_dda(i,i)   =  (1+dp*g[i]*theta)*Stensor4::Id()+C[i]*dp*theta*iseq/1.5*Jmn;
    }
    dfa_dda(0,1)   =  2*C[1]*dp*theta*iseq/3*Jmn;
    dfa_dda(1,0)   =  2*C[0]*dp*theta*iseq/3*Jmn;
  }
}

**Comment by thelfer** [comment]
Hi @GhiathMonnet,

Your question are related to the StandardElasticity brick.

This brick is described here, including the `computeElasticPrediction` method and the fact that `feel` is initialized to `deel-deto` before the evaluation of the residual in `@Integrator` : https://thelfer.github.io/tfel/web/HookeStressPotential.html

**Comment by GhiathMonnet** [comment]
Hi @thelfer,
many thanks for the answers. I understand why it works now (:): 
I read (https://thelfer.github.io/tfel/web/HookeStressPotential.html) several times and I was not able to understand it, no matter my guesses on the possible meanings of the different sentences. For example : 
1- what the mfront user has to do to declare or use the "Hooke stress potential" ?  
2- it is said that (The total strain increment is automatically substracted to the equation associated with the elastic strain : feel -= deto). I saw different implementations (see Norton law) where the user had to explicitely write (  feel += dp*n - deto ;) in the integration block. So I do not understand the word "automatically", 
3- when it is said (the stiffness tensor must be available), what that means ? Available for who and by who ? Is it possible to express this sentence differently ? How a Hooke-based law can work if the stiffness tensor is not available ?    
4- it is said that (The Hooke stress potential automatically defines the computeElasticPrediction mehod). What that mean "automatically defines" ? Does that mean, the Hooke stress potential (although it is not clear how to invoke it) applies the method If it is automatic, why it is explicitly called in the Chaboche implementation ?  
5-Anyway, I searched this page. It is said that the computeElasticPrediction computes a prediction of the stress .... except the elastic strain which are assumed to be equal to eel +deto). Did this mean that the stress is computed using the elastic state (eel + deto), i.e. without changing the value of the variable (eel)  or that eel has been explicitly updated by adding deto ? 

Thank you very much for your vital help.
Ghiath

**Reply by thelfer** [reply]
Maybe the documentation is misleading. The key point is (https://thelfer.github.io/tfel/web/BehaviourBricks.html): 

> the StandardElasticity brick, which has been described in the introduction of this page.
> In practice, the evolution of MFront and the development of the StandardElastoViscplasticity brick, has made the
 > StandardElasticity brick a kind of special case of the StandardElastoViscoplasticity where the stress potential is
>  limited to the Hooke’ one and no inelastic flow is allowed. As such, most information concerning
>  the StandardElasticity brick can be found on the page dedicated to the Hooke stress potential. In particular,
> all options passed to the StandardElasticity brick are indeed forwarded to the underlying Hooke stress potential.

Hence, you use the `HookeStressPotential` as soon as you use the `StandardElasticity` brick.

1.  See previous remark
2. Those implementations do not use the `StandardElasticity` brick.
3. The stiffness tensor must be available: it means that the law must provide a way to compute the stiffness tensor. For orthotropic behaviors, you may rely on keywords like `@RequireStiffnessTensor`,  `@ComputeStiffnessTensor` or `@ElasticMaterialProperties`. For isotropic behaviors, you can use the previous keywords, but you can also simply define the elastic material properties using either `@MaterialProperty` or `@Parameter`: in this case, the associated variables must have the glossary names `YoungModulus` and `PoissonRatio`.
4. If you use the `StandardElasticity` brick, the `computeElasticPrediction` is automatically **generated**, so that you can use it if you need it
5. It means that we compute `D:(eel+deto)` where `eel` is the value of the elastic strain at the beginning of the time step. This value is not modified.

**Comment by GhiathMonnet** [comment]
Hi Thomas and everybody,
I am trying to use @Brick "StandardElasticity and still do not understand what this brick does exactly in mfront. I checked several times the (https://thelfer.github.io/tfel/web/HookeStressPotential.html) page and nothing is said about this  brick :
-  at which moment the stress tensor sig is updated by the elastic prediction ? Is this already done in the @InitLocalVariables  block  or in the @Predictor block ? 
- what is the effect of this brick on the computation of the Jacobian ? 
- In the integrator block, can we compute the intermediate vale of seq = D * (eel + theta * deel) ?
- what is the effect of this brick on the calculation of the consistent tangent operator ?
Many thanks and hope you a nice wk.
Cheers
 Ghiath

**Reply by thelfer** [reply]
Hi @GhiathMonnet,

> at which moment the stress tensor sig is updated by the elastic prediction ? Is this already done in the @InitLocalVariables block or in the @Predictor block ?

> Computation of the elastic prediction of the stress

The page that you cited has this quotation:

> The Hooke stress potential automatically defines the computeElasticPrediction method which computes a prediction of the
> stress under the assumption that all states variables are equal to their values at the beginning of the time step except the
> elastic strain which are assumed to be equal to  
 
 In practice, you have to call `computeElasticPrediction` explicitly, probably in the `@Predictor` block:
 
```c++
@Predictor{
   const auto sigel = computeElasticPrediction();
}
``` 

The stress tensor is always initialized to its value at the beginning of the time step. The brick `StandardElasticity` does not change that.

> what is the effect of this brick on the computation of the Jacobian ?

It has no effect. The default values of the jacobian is the same as in the `Implicit` DSL.

However, the   `StandardElasticity`  does have an effect of the residual, since `feel` is initialized to `deel-deto` and not `deel`.

 > In the integrator block, can we compute the intermediate vale of seq = D * (eel + theta * deel) ?
 
 This is done by the @ComputeStress block which is automatically defined by the brick `StandardElasticity` .

Thus, in the `@Integrator` block, `sig` has the value `D * (eel + theta * deel)`. See also this page on how the blocks are called in the Implicit DSL: https://thelfer.github.io/tfel/web/implicit-dsl.html

> what is the effect of this brick on the calculation of the consistent tangent operator ?

The consistent tangent operator is automatically computed following the rules given here: https://thelfer.github.io/tfel/web/implicit-dsl.html#sec:tangent_operator:small_strain_behaviours

**Reply by GhiathMonnet** [reply]
Thank you Thomas. 
I am still somehow confused by your answer :
- you say "The stress tensor is always initialized to its value at the beginning of the time step." but then you also say "in the @Integrator block, sig has the value D * (eel + theta * deel). "  ?
- I still do not understand : if I am using the Brick, do I have to use computeElasticPrediction() or it is automatically done ?
- my question concerned @Brick StandardElasticity and cited "The Hooke stress potential ...." Do you mean that @Brick StandardElasticity does everything described in the https://thelfer.github.io/tfel/web/HookeStressPotential.html ? 
- 
- I saw different mfront files using @Brick StandardElasticity without calling computeElasticPrediction(). Does this mean that  sig = D * eel or sig = D * (eel + theta * deel) ? Does this change between the @Predictor and @integrator blocks ?
- Let's say I am only using the @Brick StandElasticity (without calling other instructions such computeElastic...) . does the value of sig change between the @Predictor and @integrator blocks ?

**Reply by thelfer** [reply]
> you say "The stress tensor is always initialized to its value at the beginning of the time step." but then you also say "in the @Integrator block, sig has the value D * (eel + theta * deel). " ?

It means that `sig` will have the value at the beginning of the time step in `@InitLocalVariables` and `@Predictor`. During the implicit algorithm, it will be equal to `D * (eel + theta * deel)`.

After the implicit algorithm, notably in `@UpdateAuxiliaryStateVariables`, it will be equal to `D * eel` (after the update of `eel` which now has the value at the end of the time step).

> I saw different mfront files using @brick StandardElasticity without calling computeElasticPrediction(). Does this mean that sig = D * eel or sig = D * (eel + theta * deel) ? Does this change between the @Predictor and @Integrator blocks ?

Well, if the behaviour does not need to estimate the elastic prediction, `computeElasticPrediction` is not called. Generally,  `computeElasticPrediction`  is not used to update the `sig` variable. See my example, where I initialized a (new) variable named `sigel`.

> Let's say I am only using the @brick StandElasticity (without calling other instructions such computeElastic...) . does the value of sig change between the @Predictor and @Integrator blocks 

Yes, see the first point

**Comment by GhiathMonnet** [comment]
OK Thank you very much Thomas. Now it is clear in my head :+1: 
Hope You a nice day.
Cheers
